Introduce early print

povergoing opened the pull request:
Introduce an early print `EARLY_PRINT` which indicates that the CPU can do printk or customized print at the very beginning stage. Introduce a Kconfig choice to support print backend selection in multi serial device cases.

Also, add the necessary function stubs so that the uart driver can easily support such a feature.
Enable early uart by implementing several necessary functions. 

Taking Arm64 as the first arch to support such a feature, handled the uart accesses with SMP in both mapped and unmapped cases.
Use `menuconfig` or `CONFIG_EARLY_PRINT=y CONFIG_EARLY_PRINT_USE_UART_PL011=y` to enable early print with pl011 uart on Arm64 platforms.

Tested with `qemu_cortex_a53`, `qemu_cortex_a53_smp`, `fvp_baser_aemv8r`, and `fvp_baser_aemv8r_smp`

NOTE: This is not an early console as there is already an [early console](https://github.com/zephyrproject-rtos/zephyr/blob/934a09550dd17406426ac92aac89761cbe506520/subsys/debug/Kconfig#L223). This is just an early print with limited features aiming at debugging at early boot stage (e.g. before `z_cstart()`)

carlocaione commented:
I like this very much. Two things though:
1. Why this is `ARM64` specific? This could be very well a generic API like `earlycon` on Linux
2. Do we want to keep this in the `serial` API? Maybe could be worth it to introduce an entirely and separate API for this?

SgrrZhf commented on drivers/serial/uart_pl011.c:
```diff
@@ -286,14 +290,16 @@ static const struct uart_driver_api pl011_driver_api = {
 	int ret;
 	uint32_t lcrh;
 
+#ifndef CONFIG_ARM_VERY_EARLY_CONSOLE
 	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
+#endif
```
When multiple pl011 devices are enabled in the devicetree, others are not mapped except the one the console selects.

povergoing replied:
Good point, I need to fix it.

povergoing commented:
Really nice suggestion

> Why this is ARM64 specific? This could be very well a generic API like earlycon on Linux
> Do we want to keep this in the serial API? Maybe could be worth it to introduce an entirely and separate API for this?

I was thinking of doing something like earlycon which is apparently a larger work. I am not sure if this is really needed in Zephyr at that time. Now I am going to take a look and try to introduce a more generic API to do this since you think this is useful. :)

povergoing committed:
subsystems: debug: Introduce early print

Introduce an early print EARLY_PRINT which indicates that the CPU can do
printk or customized print at the very beginning stage. Introduce a
Kconfig choice to support print backend selection in multi serial device
cases.
Also, add the necessary function stubs so that the uart driver can
easily support such a feature.

Signed-off-by: Jaxson Han <jaxson.han@arm.com>

povergoing committed:
drivers: serial: uart_pl011: Enable early uart

Enable early uart by implementing several necessary functions. Taking
Arm64 as the first arch to support such a feature, handled the uart
accesses with SMP in both mapped and unmapped cases.

Signed-off-by: Jaxson Han <jaxson.han@arm.com>

carlocaione commented on drivers/serial/Kconfig.earlyprint:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2023 Arm Limited (or its affiliates). All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Configs list here will be included into EARLY_PRINT, see subsys/debug/Kconfig,
+# to support early print mode choices.
+config EARLY_PRINT_USE_UART_PL011
+	bool "Use uart pl011"
+	depends on UART_PL011
+	help
+	  Use uart pl011 to do early print
```
```suggestion
# Configs list here will be included into EARLY_PRINT, see subsys/debug/Kconfig,
# to support early print mode choices.

config EARLY_PRINT_USE_UART_PL011
	bool "Use UART PL011"
	depends on UART_PL011
	help
	  Use UART PL011 to do early print
```

carlocaione replied:
Also this should go in the next commit.

povergoing replied:
Thanks, will fix it.

carlocaione commented on include/zephyr/debug/early_print.h:
```diff
@@ -0,0 +1,26 @@
+extern "C" {
+#endif
+
+/*
+ * Different arch has its own demand of init stages.
+ */
+#define EARLY_PRINT_INIT_0	0
+#define EARLY_PRINT_INIT_1	1
```
Not clear to me what is the difference

povergoing replied:
While I defined two functions something like `xxx_early_print_init` and `xxx_early_map` at the first design. But I really don't know if this is generic enough regarding all architectures, not sure if there is a platform that needs 3 or more stages to init the early print, that is why I gave an unclear stage definition here. 

I can come up with two solutions if you think the specific stage name is much better:
1) change the two definitions to `EARLY_PRINT_INIT` and `EARLY_PRINT_MAP`, and still use a single init function like what it is like now.
2) use two functions like `z_early_print_init` and `z_early_print_map`.
Any suggestions?

carlocaione replied:
Just add a comment here explaining that, that will be enough. Just to justify why we have two init levels.

povergoing replied:
Okay, I got it, will add some explanations

povergoing replied:
Hi Carlo, I added the comments to make it more clear. Also, change the the two definitions to EARLY_PRINT_INIT and EARLY_PRINT_MMIO_MAP, as I think it might be generic enough for all MMU system.

carlocaione commented on drivers/serial/uart_pl011.c:
```diff
@@ -366,6 +376,81 @@ static int pl011_init(const struct device *dev)
+#endif
+}
+
+int arch_printk_char_out(int c)
```
Is this function supposed to be used by code? Who is the user?

povergoing replied:
Oh, it's the `printk`'s default char out hook, https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/os/printk.c#L46-L55

carlocaione replied:
oh, TIL.

nashif commented on subsys/debug/Kconfig:
```diff
@@ -228,6 +228,23 @@ config EARLY_CONSOLE
 	  initialize its driver earlier than normal, in order to get the stdout
 	  sent through the console at the earliest stage possible.
 
+config EARLY_PRINT
```
We already have CONFIG_EARLY_CONSOLE, why can't this be used? This looks very custom to a use case and we need to generalize this

povergoing replied:
The `CONFIG_EARLY_CONSOLE` focuses on `input` and `output` at a early boot stage but cannot be earlier than interrupt controller init. The `CONFIG_EARLY_CONSOLE` is initialized after z_cstart(). However, when developers want to bring up a new platform, they probably need to print something (only output) for debugging even before interrupt controller or MMU. I found x86 also has this kind of early print https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/x86/core/early_serial.c), we need a similar one for Arm64, that's the main reason I want to introduce this but this is more generic.

nashif replied:
> I found x86 also has this kind of early print [`main`/arch/x86/core/early_serial.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/x86/core/early_serial.c?rgh-link-date=2023-11-29T02%3A07%3A35Z)), we need a similar one for Arm64, that's the main reason I want to introduce this but this is more generic.

You are right, I was actually thinking about this one which is more lile what you are looking for and it is controlled with CONFIG_X86_VERY_EARLY_CONSOLE which pretty much is the same as EARLY_PRINT. 

We need to generalize this, it does not make sense to have different architecture have different ways for doing the same thing, I do not think any of this supports input at this stage, even if they have CONSOLE in the name.

povergoing replied:
Okay Anas, if you prefer, I could change the `X86_VERY_EARLY_CONSOLE`, and make it under the `EARLY_PRINT`?

nashif replied:
wait on that, let me think about it a bit more, @jhedberg  any input on this?

povergoing replied:
Sure, take your time

jhedberg replied:
Yeah, at first sight it looks like the `X86_VERY_EARLY_CONSOLE` functionality should be moved behind this new more generic Kconfig option.

povergoing replied:
Can I ask if there are any more suggestions?

nashif commented on include/zephyr/debug/early_print.h:
```diff
@@ -0,0 +1,30 @@
+#define EARLY_PRINT_INIT	0
+#define EARLY_PRINT_MMIO_MAP	1
+
+void z_early_print_init(int stage);
```
is this a public API? We should not use z_ prefix for public APIs, also needs proper doxygen docs

povergoing replied:
Yes, this is a public API. I can change it to `void early_print_init(int stage)`, is this name appropriate? And yes, I will add the docs
